### ISSUE TITLE
no-irregular-whitespace: Inside of regexp irregular whitespace

### ISSUE THREAD

**Issue by inuscript** [eslint/eslint#5840]
**What version of ESLint are you using?**
2.7.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```
{
    "env": {
        "node": true,
    },
    "rules": {
        "no-irregular-whitespace": 2
    }
};
```
**What did you do? Please include the actual source code causing the issue.**

I wrote those code.

```js
var reg = /^[\s\　]*$/
reg.test('foo')
```

**What did you expect to happen?**

```
  1:5  error  Parsing error: Unexpected token reg
```

I think this may bug because string literal is ignored on this rule.

If this is not bug, I want documentation or `skipRegexp` option.

Note: I can avoid this error with `new RegExp('^[\s\　]*$')`

**Comment by mysticatea** [comment]
Thank you for this issue.

Indeed, `no-irregular-whitespace` ignores spaces in string literals.
So I feel this is a bug.

![image](https://cloud.githubusercontent.com/assets/1937871/14513794/635ce0a0-0225-11e6-85c7-26a718998525.png)

**Comment by kaicataldo** [comment]
I actually don't think this is a bug - seems to be working as intended, as the rule states it causes warnings if irregular whitespace characters are found outside of strings or comments. From the source: 
> With this rule enabled the following characters will cause warnings outside of strings and comments ...

I think that it's not working for your first example because it's a regex literal (not a string). The reason `new RegExp('^[\s\　]*$')` doesn't warn is because the irregular whitespace is in a string. Or that's my interpretation of the rules docs, anyway :)

**Comment by inuscript** [comment]
Oh sorry, my explanation was lacking.

If this is regular behavior, I think this should documentation with incorrect example.
But I wish fix this behavior or append ignore option.

Because regexp with `Ideographic Space`' pattern is often used in multi byte language.

**Comment by kaicataldo** [comment]
Got it - I'm going to label this as an enhancement and let's see what @nzakas thinks. Thanks for the issue report!

**Comment by mysticatea** [comment]
Hmm, I cannot find reasons that irregular spaces in strings are good but irregular spaces in regular expressions are bad. And I got aware that template literals have the same issue:

```js
/^[\s\　]*$/             /*error Irregular whitespace not allowed (no-irregular-whitespace)*/
new RegExp(`^[\s\　]*$`) /*error Irregular whitespace not allowed (no-irregular-whitespace)*/
new RegExp("^[\s\　]*$") // good.
```

I wish to fix this behavior since this does not have consistency, IMO.

What about adding options for each type?

```js
{
    "no-irregular-whitespace": ["error", {
        "skipStrings": false,
        "skipTemplates": false,
        "skipRegExps": false,
        "skipComments": false // "skipComments" exists already.
    }]
}
```

In my ideal, all default are `false`, but it's a breaking change.

**Comment by nzakas** [comment]
Irregular whitespace in strings might be intentional for output. No objections to adding the options if @mysticatea wants to as long as the default behavior remains the same.

**Comment by rhysd** [comment]
I just encountered this bug today and sent a PR for this in #6018.  Could you check it?

**Comment by mysticatea** [comment]
@rhysd Thank you for the PR :smile:. But the consensus on this issue is to add options above. Could you update the PR with the options?

**Comment by rhysd** [comment]
@mysticatea I'll fix my PR today :+1:

**Comment by rhysd** [comment]
```javascript
        "skipStrings": false,
        "skipTemplates": false,
```

Should we separate these options?  I think `skipTemplates` should be unified into `skipStrings`.  What do you think?

**Comment by pedrottimark** [comment]
@rhysd If we were designing the rule today, unlikely to find a use case to separate options for string literals from template literals. The reason is to avoid breaking existing behavior, as requested in https://github.com/eslint/eslint/issues/5840#issuecomment-211439304.

Therefore, I think the default values of the options are as follows: 

```js
{
    "no-irregular-whitespace": ["error", {
        "skipStrings": true, // current default is to skip strings
        "skipTemplates": false,
        "skipRegExps": false,
        "skipComments": false // this option already exists
    }]
}
```

**Comment by rhysd** [comment]
@pedrottimark 

OK, I'll maintain the previous behavior :+1:
